### PR TITLE
Tdr 2558 update view transfers with links to pages

### PR DESCRIPTION
**Pull request by techncl**
This is only for the standard flow; we can do the judgment flow in another PR

### CODE REVIEW

**Review by TomJKing — COMMENTED** [review]
Is there a reason why did do a match on the current status, like the pseudo code below:

  ```
private def toStandardUserAction(statuses: CurrentStatus): Unit = {
    statuses match {
      case s if s.series.isEmpty => UserAction("/series", "Resume Transfer")
      case s if s.series.nonEmpty && s.series.get == "Completed" => UserAction("/transfer-agreement", "Resume Transfer")
      case s if s.transferAgreement.nonEmpty && s.transferAgreement.get == "InProgress" => UserAction("/transfer-agreement-continued", "Resume Transfer")
      case s if s.transferAgreement.contains("Completed") && s.clientChecks.isEmpty => UserAction("/upload", "Resume Transfer")
      case s if s.clientChecks.contains("InProgress") => UserAction("/file-checks ", "Resume Transfer")
      case s if s.clientChecks.contains("CompletedWithIssues") => UserAction("/file-checks-results", "Failed")
      case s if s.clientChecks.contains("Completed") && s.confirmTransfer.isEmpty => UserAction("/file-checks-results", "Resume Transfer")
... etc ...
      case _ => UserAction("/contact-us", "Contact us")
    }
  }
```
Doing it like this, keeps the logic of what link/text to display in one place, which is the most complicated part of the feature. The default would always be the "Contact us" action, which would still allow the user to a least see the consignments, even if the statuses get changed in the DB.

**Comment by techncl** [comment]
> Is there a reason why did do a match on the current status, like the pseudo code below:
> 
> ```
> private def toStandardUserAction(statuses: CurrentStatus): Unit = {
>   statuses match {
>     case s if s.series.isEmpty => UserAction("/series", "Resume Transfer")
>     case s if s.series.nonEmpty && s.series.get == "Completed" => UserAction("/transfer-agreement", "Resume Transfer")
>     case s if s.transferAgreement.nonEmpty && s.transferAgreement.get == "InProgress" => UserAction("/transfer-agreement-continued", "Resume Transfer")
>     case s if s.transferAgreement.contains("Completed") && s.clientChecks.isEmpty => UserAction("/upload", "Resume Transfer")
>     case s if s.clientChecks.contains("InProgress") => UserAction("/file-checks ", "Resume Transfer")
>     case s if s.clientChecks.contains("CompletedWithIssues") => UserAction("/file-checks-results", "Failed")
>     case s if s.clientChecks.contains("Completed") && s.confirmTransfer.isEmpty => UserAction("/file-checks-results", "Resume Transfer")
> ... etc ...
>     case _ => UserAction("/contact-us", "Contact us")
>   }
> }
> ```
> 
> Doing it like this, keeps the logic of what link/text to display in one place, which is the most complicated part of the feature. The default would always be the "Contact us" action, which would still allow the user to a least see the consignments, even if the statuses get changed in the DB.

Ah, yes, this is a better idea; I think because I had originally started in the ConsignmentStatusService so I had iterating through a list of statuses on my mind but didn't rethink when I moved the code to the frontend. Thanks

File: app/controllers/ViewTransfersController.scala [hunk 1]
```diff
@@ -13,8 +13,36 @@ import java.time.format.DateTimeFormatter
...
+// deliberately not using .get for Maps so that it throws an exception as we should be alerted if values in DB have changed
 class ViewTransfersController @Inject() (val consignmentService: ConsignmentService, val keycloakConfiguration: KeycloakConfiguration, val controllerComponents: SecurityComponents)
     extends TokenSecurity {
+
+  private val inProgressStepStatus = "InProgress"
+  private val failedStepStatus = "Failed"
+  private val completedWithIssuesStepStatus = "CompletedWithIssues"
+  private val completedStepStatus = "Completed"
+
+  private val inProgressConsignment = "In Progress"
```
Comment (TomJKing) [inline app/controllers/ViewTransfersController.scala:25]: I would rename these like this `transferStatusInProgress` etc, then it matches with the name of the field in the `UserAction` case class

File: app/controllers/ViewTransfersController.scala [hunk 2]
```diff
@@ -13,8 +13,36 @@ import java.time.format.DateTimeFormatter
 import java.util.UUID
 import javax.inject.Inject
 
+// deliberately not using .get for Maps so that it throws an exception as we should be alerted if values in DB have changed
 class ViewTransfersController @Inject() (val consignmentService: ConsignmentService, val keycloakConfiguration: KeycloakConfiguration, val controllerComponents: SecurityComponents)
     extends TokenSecurity {
+
+  private val inProgressStepStatus = "InProgress"
```
Comment (TomJKing) [inline app/controllers/ViewTransfersController.scala:20]: I'd remove the `Step` from the name of these.

File: test/controllers/ViewTransfersControllerSpec.scala [hunk 3]
```diff
@@ -135,16 +180,42 @@ class ViewTransfersControllerSpec extends FrontEndTestHelper {
...
   }
+
+  private val getStatusColour = Map(
+    "In Progress" -> "yellow",
+    "Failed" -> "red",
+    "Transferred" -> "green",
+    "Contact us" -> "red"
+  )
+
+  private val allPossibleStatusesAndActionUrlsForStandardUsers: List[(String, String)] = List(
```
Comment (TomJKing) [inline test/controllers/ViewTransfersControllerSpec.scala:205]: Maybe rename name this: `standardStatusesAndActionUrls`

File: test/testUtils/FrontEndTestHelper.scala [hunk 4]
```diff
@@ -729,6 +734,35 @@ trait FrontEndTestHelper extends PlaySpec with MockitoSugar with Injecting with
     "Failed"
   )
 
+  val standardConsignmentStatusValuesThatGenerateActionLinks: Map[String, List[Option[String]]] =
+    ListMap(
+      // None = user has not started transfer (go to series page)
```
Comment (TomJKing) [inline test/testUtils/FrontEndTestHelper.scala:739]: I wouldn't add all these comments in here. Comments have a tendency not to get updated, and become incorrect over time, causing more confusion than they actually solve.

This links with my previous comment of writing specific unit tests for the `convertStandardStatusesToUserActions` function. By having these it makes clear what the expected behaviours are, without the need for the comments.

File: app/controllers/ViewTransfersController.scala [hunk 5]
```diff
@@ -24,28 +52,74 @@ class ViewTransfersController @Inject() (val consignmentService: ConsignmentServ
...
-    )
-  }
+    }
+
+  private val formatter = DateTimeFormatter.ofPattern("dd/MM/yyyy HH:mm:ss")
 
-  private def getConsignmentStatus(currentStatus: CurrentStatus): String = {
-    if (currentStatus.`export`.contains("Completed")) "Exported" else "In Progress"
+  // scalastyle:off cyclomatic.complexity
+  private def convertStandardStatusesToUserActions(statuses: CurrentStatus, consignmentId: UUID, consignmentRef: String) = {
```
Comment (TomJKing) [inline app/controllers/ViewTransfersController.scala:79]: I would actually not make this private, so you can write specific unit tests for this function. The logic is complex, and it justifies having unit tests for it, to help developers in the future understand the expected behaviours clearly.

File: test/controllers/ViewTransfersControllerSpec.scala [hunk 6]
```diff
@@ -109,17 +130,41 @@ class ViewTransfersControllerSpec extends FrontEndTestHelper {
...
+      consignmentType: String,
+      node: Node,
+      index: Int,
+      statusesAndActionUrls: List[(String, String)] = allPossibleStatusesAndActionUrlsForStandardUsers
+  ): Unit = {
     val exportDate = node.exportDatetime.map(_.format(formatter)).get
     val createdDate = node.createdDatetime.map(_.format(formatter)).get
-    val status = if (node.currentStatus.`export`.contains("Completed")) "Exported" else "InProgress"
+    val (transferStatus, actionUrlWithNoConsignmentId): (String, String) =
+      if (consignmentType == "judgment") ("", "") else statusesAndActionUrls(index) // need to implement one for judgments
```
Comment (TomJKing) [inline test/controllers/ViewTransfersControllerSpec.scala:160]: Should be implementing the feature for judgments, and testing as well, even though the feature won't be released for judgment users just yet.

**Review by TomJKing — COMMENTED** [review]
Looks good. Just some comments. Main thing is should be implementing for judgments aswell, even though the feature might not be released for judgment users just yet.

**Review by TomJKing — APPROVED** [review]
Looks good. Thanks

**Pull request merged**